java/core/src/main/java/sleeper/core/util/ByteArray.java

```diff
@@ -0,0 +1,206 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+/*
+ * Copyright 2022-2025 Crown Copyright
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package sleeper.core.util;
+
+import java.util.Arrays;
+import java.util.Comparator;
+
+/**
+ * Code utilised below is a subset of the ByteArray class from facebook.collections.
+ *
+ * Class to provide wrapper object for primitive byte array into an object.
+ * Require as hashCode and equals requied for usage of type within a HashSet.
+ *
+ */
+public abstract class ByteArray implements Comparable<ByteArray> {
+
+    public static final ByteArrayComparator BYTE_ARRAY_COMPARATOR = new ByteArrayComparator();
+
+    /**
+     * Returns contents of object as primitive type.
+     *
+     * @return array details in primitive type
+     */
+    public abstract byte[] getArray();
+
+    /**
+     * Returns length of the array within the object.
+     *
+     * @return length of array
+     */
+    public abstract int getLength();
+
+    /**
+     * Simple method for wrapping primitive type within an object.
+     *
+     * @param  array primitive byte array
+     * @return       wrapped object
+     */
+    public static ByteArray wrap(byte[] array) {
+        return new PureByteArray(array);
+    }
+
+    /**
+     * Method for checking equaility.
+     *
+     * @param  array1 first ByteArray object
+     * @param  array2 second ByteArray object
+     * @return        true if equal
+     */
+    public static boolean equals(ByteArray array1, ByteArray array2) {
+        //If both null, then they are equal
+        if (array1 == null && array2 == null) {
+            return true;
+        }
+        //If one is null, then equality is false
+        if (array1 == null || array2 == null) {
+            return false;
+        }
+        // If set, proceed to check via arrays methodology
+        return Arrays.equals(array1.getArray(), array2.getArray());
+    }
+
+    /**
+     * Method for checking equality versus generic object declaration.
+     *
+     * @param  o object to compare versus
+     * @return   true if equal
+     */
+    public boolean equals(Object o) {
```

Comment:
Usually for equals/hashCode/toString we use an `@Override` annotation and inherit the Javadoc from Object. The Javadoc you get when you just use `@Override` explains a bit more.

There are a number of instances of this in this file.